### DISCUSSION TITLE
Joomla 4.0

### DISCUSSION THREAD

**Discussion by LordRazen** [joomengine/Joomla-Component-Builder#649] [Extention Issues] [answered]
I just did a bit research on Joomla 4.0 and saw the release candidate for 4.0 is planned for february. 

Since I think about creating a component with component builder, I wonder what the plan is here. 

I've no idea how much effort it will be to update component builder or if you could prepare this already. I also wonder if the components made by component builder will be compatible or we wil probably face heavy issues. 

If you would let me know a few insights about the update process for component builder on Joomla 4.0 and maybe a few words on your estimated time for the update process, it would be awesome. Every info is interesting for me ^^

Thanks :)

**Comment by Llewellynvdm** [comment, marked as answer]
We are aiming to have JCB write Joomla 4 native components by end of **July 2021** if all goes according to plan.

- That means that JCB itself will run on Joomla 4 and build for Joomla 4 and be able to migrate/move all your components to Joomla 4.
- Do take note that we do not convert any custom code you may have added, via PHP, JS, and/or CSS areas across JCB. You will have to manually update those where needed.

The path to achieve this (moving JCB to J4) is not that hard, we can already build components that suppose to work in Joomla 4 (legacy mode, which is not fully ready in J4) Since J3.10 is the J4 legacy, and JCB runs easily on J3.10

But I would like to make the needed changes to ensure JCB is native J4 by the end if July 2021.

The compiler may **look daunting** since it breaks all common design patterns/practice. But to me it is a **very simple** and **straightforward** solution really. When I think about the compiler it is a very smart array **type** interface system (which works with a hash table like approach), that require very little resources to understand/build what you told it to do.

Much conversation with @nibra, @marcodings, and other has strongly pointed me to refactoring on the compiler to a much more common design pattern and trusted practices. But I have been very over loaded with work, and at the same time thinking that since it works, why fix it. Lets face it, many advance developers do not want to take the time to use JCB (which is sad, and I really hope that will change), they like to write their own code and use their own tools (which is often not public). Much more when they see the compiler they are even more reluctant to contribute, since they can't easily understand what I did. Which is clearly considered bad practice, and I would agree most of the time.

But with JCB I would say, building a tool that does what JCB does, and that on almost any shared hosting environment... I had to follow a path that is not that ideal, but is now able to do what everyone said can't be done. Since JCB does not just build components, it builds advance and highly complex components... and that with nearly no limitation, and that extremely accurate.

The fact that I have maintained more then 50 components for more then 3 years in use by various clients with JCB with literally no bugs, besides those I may have cause with my own additional custom code I added in JCB. I mean just look at [SD](https://github.com/Llewellynvdm/Joomla-Sermon-Distributor), and the [large amount of changes](https://github.com/Llewellynvdm/Joomla-Sermon-Distributor/graphs/contributors?from=2018-12-27&to=2021-01-16&type=c) made over the last year to the code base, due to improvements to JCB, and you will see a maintainability that most can only dream about, while I hardly ever looked at SD directly.

I manage all my components in a git repository, and when I make changes to JCB, I compile JCB and one or two other components. Then very quickly I am able to check if the changes were correct, and very easily see if all use cases are covered. To me JCB has been noting less then a tremendous time saver, and an empowering tool to achieve what would otherwise never have been possible,

So here is the moment of truth, can JCB step over to J4, and what will the work load look like. So the little hint... Each version of Joomla from 3 upwards will have its own set of templates, folder structures, and dynamic code. The whole idea of JCB was to follow Joomla, without the huge work load of manually updating the code as changes happen in Joomla. That has worked the whole path of J3, and to this day we have been making changes and taking steps to stay J3 native. I believe we are seeing the same to be true, when stepping over to another version.

**Reply by kevincam3** [reply]
So did we make it to get JCB to build Joomla 4 extensions?

**Reply by Llewellynvdm** [reply]
I am still hard at work on this... :+1: but soon....

**Comment by TLWebdesign** [comment]
Is there anything we can do to help?

**Reply by Llewellynvdm** [reply]
Nothing just yet, but once I start push the new branch we would need testers, and contributors to insure stability as soon as possible. I will bump this tread at that time.

**Comment by LordRazen** [comment]
Thanks for the detailed reply. Glad you plan to upgrade this awesome component. 

JCB also allows non-developers to create components: I know a bit PHP, but I'm miles away from creating an own Joomla component, especially since I don't know Joomlas way of doing stuff perfectly. With JCB, I can do create own components, see where the code goes and understand a lot more.

So: Being able to create own components and the possibility to do this in Joomla 4.0 is great. 

A big thanks to you and all contributors for your awesome project!
